Implement `repository_ctx.read()` for reading local files.

jmillikin opened the pull request:
Fixes #3766

dslomov commented:
Let's discuss the use-case for this in #3766.

cgruber commented on src/main/java/com/google/devtools/build/lib/bazel/repository/skylark/SkylarkRepositoryContext.java:
```diff
@@ -262,6 +263,29 @@ public void createFileFromTemplate(
+    SkylarkPath p = getPath("read()", path);
+    Charset charset;
+    if (encoding.equals("")) {
+      charset = StandardCharsets.ISO_8859_1;
```
We want 8859-1 as the default?

jmillikin replied:
The default is set to UTF-8 in `SkylarkRepositoryContextApi`. I'm using `StandardCharsets.ISO_8859_1` to mean raw bytes, for when the user calls `ctx.read("somefile", encoding="")`.

My assumption is that this function will almost always be used to ingest text files in ASCII / UTF-8, with `encoding = ""` available as an escape hatch for odd cases.

cgruber replied:
Oh, I see!  Ok.  Gotcha.

jmillikin-stripe commented:
@dslomov: ping

dslomov commented on src/main/java/com/google/devtools/build/lib/bazel/repository/skylark/SkylarkRepositoryContext.java:
```diff
@@ -262,6 +263,29 @@ public void createFileFromTemplate(
+            p.toString(), encoding, rule.getLabel().toString(), location);
+    env.getListener().post(w);
+    try {
+      return FileSystemUtils.readContent(p.getPath(), charset);
```
Starlark strings are raw bytes (https://github.com/bazelbuild/starlark/blob/master/spec.md#strings). Therefore this method should not have any encoding passed in.

jmillikin replied:
`repository_ctx.file()` assumes strings are UTF-8:
```java
public void createFile(Object path, String content, Boolean executable, Location location) {
      ...
      try (OutputStream stream = p.getPath().getOutputStream()) {
        stream.write(content.getBytes(StandardCharsets.UTF_8));
      }
```

If the encoding does not default to UTF-8, then `ctx.file("foo", s); ctx.read("foo") == s` will be false.

If we do not allow the user to specify no encoding, then it won't be possible to properly read a binary file.

If you're certain that `ctx.read()` should not have an encoding parameter, should `ctx.file()` also be changed to treat its input as ISO-8859-1 (i.e. plain bytes)?

dslomov replied:
> repository_ctx.file() assumes strings are UTF-8:

That's a bug in `ctx.file()`

> If we do not allow the user to specify no encoding, then it won't be possible to properly read a binary file.

We should read in ISO-8859-1 - then we get raw bytes.

> If you're certain that ctx.read() should not have an encoding parameter, should ctx.file() also be changed to treat its input as ISO-8859-1 (i.e. plain bytes)?

Yes (this is an incompatible change, so the way to do it is to add `legacy_utf8` boolean to `ctx.file`, defaulting to  `true` and go through incompatible change process to make the default `false`.

cgruber replied:
Or rename `ctx.file()` to `ctx.write()` to be symmetrical with `read()` and make these paired functions behave in the new way, and deprecate `file()`

Still requires the incompatible change process to deal with the deprecation, but it's probably more instructive and better named.

jmillikin replied:
OK, changes made in new commit:
* Remove `encoding=`, and reads be decoded as ISO-8859-1.
* Add `ctx.write_file()`, which is like `ctx.file()` but encodes with ISO-8859-1.
  * I also had to use a new `workspace_log.proto` message type, because the old `FileEvent` has its contents in `string` instead of `bytes`.
* Renamed new function to `ctx.read_file()` to match.

dslomov replied:
Let's keep the `ctx.file()` and add `legacy_utf8` as I suggested. Why? Migration will be no-op for most (I expect 100% of) people, where with function rename people will have to modify their code.

Let's keep the name `ctx.read()` the function you are adding.

I am not happy about the new event in BEP. Add bytes to `FileEvent` instead.

jmillikin replied:
@dslomov Above changes applied.

dslomov commented:
@c-parsons please take a look from Starlark API perspective

jmillikin commented:
(rebased to current master HEAD)

jmillikin commented:
Latest commit addresses review feedback. PTAL.

jmillikin commented:
CI is failing due to what looks like a bazel mirror outage.

dslomov commented on src/main/java/com/google/devtools/build/lib/bazel/debug/workspace_log.proto:
```diff
@@ -77,11 +77,17 @@ message FileEvent {
   string path = 1;
   // Content of the created file
-  string content = 2;
+  bytes content = 4;
```
Removing a field is a breaking change to the protocol.
Instead, let's have *two* fields for now: `string content` and `bytes bytes`, deprecating the first one. As part of an incompatible change to make legecy_utf8 default to false, we can set content to empty string and then remove that field.

jmillikin replied:
After further thought, I've reverted this message to only containing `string content`. The protobuf field has always been set from the `String content` directly, without encoding, so it doesn't make sense for it to change as part of this PR.

It might be reasonable to add a new field `bytes encoded_content` that differs depending on `legacy_utf8` mode, but I think that is best done separately.

dslomov commented on src/test/shell/bazel/bazel_workspaces_test.sh:
```diff
@@ -292,6 +292,24 @@ function test_file() {
   ensure_contains_exactly 'executable: true' 1
 }
 
+function test_read() {
```
Add a test with non-Latin-1 UTF-8 characters. Add a test for legacy_utf8=False

jmillikin replied:
Done and done.

jmillikin commented:
Thanks, didn't see that test earlier. I fixed up the calls to `createFile()` for the new signature, and also added a test of `readFile()`.

It passes locally. BuildKite CI seems fairly slow, so if it goes red again then I'll try again tonight.

jmillikin commented:
@dslomov Tests are green.

dslomov commented:
@brandjon had stylistic comments on the internal review - please review this PR.

brandjon commented on src/main/java/com/google/devtools/build/lib/bazel/repository/skylark/SkylarkRepositoryContext.java:
```diff
@@ -192,30 +192,36 @@ public void symlink(Object from, Object to, Location location)
   }
 
-  private void checkInOutputDirectory(SkylarkPath path) throws RepositoryFunctionException {
+  private void checkInOutputDirectory(String operation, SkylarkPath path) throws RepositoryFunctionException {
```
This really should be a bool or enum instead of a string. If you use a bool, the call sites should specify the param name in a comment, as in

```
checkInOutputDirectory(/*isRead=*/ true, ...)
```

jmillikin-stripe replied:
I think a `bool` would be inappropriate here, because there's no reason it wouldn't use more specific error messages for other checks in the future. I only copy-pasted `"write"` into existing callers to avoid significant diffs in unrelated tests.

brandjon replied:
I wasn't suggesting changing the error message for the existing case, I just wanted to avoid passing strings if it's only ever allowed to take on two possible values (e.g. to avoid subtle misspellings leading to errors). That said I see that this is only used for the error message, not to control the actual operation being done, so it's fine to leave it a string.

brandjon commented on src/test/shell/bazel/bazel_workspaces_test.sh:
```diff
@@ -292,6 +292,75 @@ function test_file() {
   ensure_contains_exactly 'executable: true' 1
 }
 
+function test_file_nonascii() {
```
What are we testing? Should this be named `test_read_file_nonascii`?

jmillikin-stripe replied:
The pre-existing test `test_file()` verifies the behavior of `repository_ctx.file()`. This test verifies the behavior of `repository_ctx.file()` when the file content contains codepoints outside the ASCII range.

brandjon replied:
Oh, fair enough. That was my lack of domain knowledge in this area showing through.

brandjon commented on src/test/java/com/google/devtools/build/lib/bazel/repository/skylark/SkylarkRepositoryContextTest.java:
```diff
@@ -145,16 +145,16 @@ public void testWhich() throws Exception {
-    context.createFile(context.path("foobar"), "", true, null);
-    context.createFile(context.path("foo/bar"), "foobar", true, null);
-    context.createFile(context.path("bar/foo/bar"), "", true, null);
+    context.createFile(context.path("foobar"), "", true, true, null);
```
These call sites can use some param name comments.

jmillikin-stripe replied:
Sure. That sounds like a reasonable cleanup that could be applied to this test file in a separate PR, or internal CL.

brandjon replied:
Granted they should've been there in the pre-existing code. I mention it because you're creating an additional parameter which makes the readability slightly worse, but I won't block this PR on it.
